# Исключения

```C++
try
{
  ...
  if(error)
    throw 4;
  ...
}
catch(int e)
{
  ...
}
catch(int *e)
{
  ...
  throw e;
}
```

## Использование исключений для new

```C++
try
{
  S *p = new S;
}
catch(...) // ловить все ошибки
{
  printf("error");
  throw;
}
```

В 64 битных системах try-catch почти бесплатный, а в 32 битных требуется ресурсов

```C++
uint div(uint a, uint b)
{
  if (b == 0)
    throw "division by zero";
  return a / b;
}
```

# RAII

```C++
int *x = new int[a];
int *y = new int[b]; // Данный код приводит к утечке памяти
```
Вариант исправления
```C++
struct C
{
  int *p;
  S(int *z) : p(z) {}
  ~S() {delete[] p;}
}

S x(new int[a]);
S y(new int[b]);
```

Или

```C++
struct C
{
  int *p;
  S(size_t s) : p(new int [s]) {}
  ~S() {delete[] p;}
}

S x(a);
S y(b);
```

# Explicit

```C++
struct S
{
  S() {}
  explicit S(int x) {} // запрет неявного каста
  operator int() {return 3;} // в обратную сторону
};

{
  S s(5);
  int x = s;
  s = 2; // не скомпилируется
}
```

```C++
struct S
{
  const char *p;
  S() : p(NULL) {}
  S(const char *a)
  {
    char *b = new char[strlen(a)+1];
    strcpy(b, a);
    p = b; // благодаря этому объект s владеет своими данными
  }
  S &operator = (const S &a)
  {
    delete[] p;
    char b = new char[strlen(a.p)+1];
    strcpy(b, a.p);
    p = b;
  }
  S(const S &a)
  {
    char b = new char[strlen(a.p)+1];
    strcpy(b, a.p);
    p = b;
  }
  ~S() {delete[] p;}
  void print() {print("%s", p);}
}
S f()
{
  char buf[25] = {};
  S s(buf);
  return s;
}
int main()
{
  S s = f();
  s = f();
  return 0;
}
```
